### ISSUE TITLE
Data Library "Link files instead of copying" Producing "File Not Found" Errors

### ISSUE THREAD

**Issue by quacksawbones** [galaxyproject/galaxy#14054]
When loading a file into a Galaxy Data Library using the "Link files instead of copying" checkbox, while the files themselves are read into the Galaxy Data Library (i.e. I can see the size of the data files and see the data within the file), but when I try to use the linked file within a tool, I get a "File Not Found" error. Also, whenever I try to download the file, my browser cannot find the file and my browser either just opens an empty page with no address bar (Chrome) or I get a "Page Not Found" error (Firefox).

This behaviour is seen regardless of who the owner is and whether the files are "Read Only" or "Read/Write" (and Galaxy would not be able to load the preview of the data file if it could nto read it in the first place).

Galaxy Version: release_22.01
Commit: 6a00e75410c7ab81d6c3d4c49a6272167e6ee6ec (Ubuntu, Ansible) and d4001d18511fa9aba30da88fce5ee98f1b1a11bb (CentOS 7)

1. Open a data library (existing or otherwise)
2. Select the "+ Datasets" button and add a data file (e.g. a FASTQ file) either "From Path" or from a permitted folder
3. Ensure the "Link files instead of copying" checkbox is ticked (i.e. True)
4. Select "Export to History" and send it to either a new History or an existing one
5. Use the data file in an analysis (for example, `Trimmomatic`, `MultiQC` or `FastQC`
6. "File not Found" errors are returned for the tool saying the data file is not there

There should be no difference in functionality between adding a data file proper and linking to it.

**Comment by quacksawbones** [comment]
I am wondering if this has something to do with the `tus` data uploader? Because if `tus` stores a temporary version of the file during upload, once the upload is complete, maybe once the temporary file is removed, the link in Galaxy is no longer valid (but the preview of the data and the reference to it in the history and data library remains, as that's stored int he database)?

**Comment by bgruening** [comment]
@quacksawbones wired. EU is using this feature a lot and we haven't seen any problems so far.

**Comment by quacksawbones** [comment]
Absolutely weird, especially seeing as I DON'T see it on previous builds! I'm wondering if the reason I'm seeing issues is because both of my builds start from release_22.01, whereas the others (like the EU server) are upgrading from a previous Galaxy release to 22.01?

**Comment by quacksawbones** [comment]
No worries. Just trying to think of what was different between 22.01 and previous releases.

Interestingly, I'm only seeing the behaviour when I load the file using the **path paste upload** method, not when I use the **library upload folder** method.

Contents of `upload_params_*` when using path paste:
```[{"file_type": "auto", "ext": null, "name": "reads_1.fq", "dataset_id": 11, "dbkey": "?", "type": "path_paste", "is_binary": null, "link_data_only": "link_to_files", "uuid": null, "to_posix_lines": true, "auto_decompress": true, "purge_source": true, "space_to_tab": false, "run_as_real_user": false, "check_content": true, "path": "/opt/data/galaxy/reads_1.fq"}]```

Contents of `upload_params_*` when using a library upload directory:
```[{"file_type": "auto", "ext": null, "name": "reads_2.fq", "dataset_id": 16, "dbkey": "?", "type": "server_dir", "is_binary": null, "link_data_only": "link_to_files", "uuid": null, "to_posix_lines": true, "auto_decompress": true, "purge_source": true, "space_to_tab": false, "run_as_real_user": false, "check_content": true, "path": "/opt/data/darrenc/reads_2.fq"}]```

**Comment by mvdbeek** [comment]
OK, still no luck reproducing this.

- Is `/opt/data/galaxy/reads_1.fq` still on disk ?
- If you click on the info button, do you see `/opt/data/galaxy/reads_1.fq` as the `Full Path` ?
- Can nginx access `/opt/data/galaxy/reads_1.fq` ?
- Is that path accessible from your cluster nodes ?

**Comment by mvdbeek** [comment]
Can you paste one of the generated command lines for a job that failed ?

**Comment by quacksawbones** [comment]
Logs attached (which includes the logs for the file upload/linking and the failed Trimmomatic job):
[galaxy_link_logs.txt](https://github.com/galaxyproject/galaxy/files/8904362/galaxy_link_logs.txt)


And the contents of the relevant `upload_params_*` file:
```[{"file_type": "auto", "ext": null, "name": "reads_2.fq", "dataset_id": 299, "dbkey": "?", "type": "path_paste", "is_binary": null, "link_data_only": "link_to_files", "uuid": null, "to_posix_lines": true, "auto_decompress": true, "purge_source": true, "space_to_tab": false, "run_as_real_user": false, "check_content": true, "path": "/hpc/software/installed/bowtie2/2.3.5.1/example/reads/reads_2.fq"}]```

**Comment by mvdbeek** [comment]
Oh, so you're running with singularity, but you're not mounting in the input path. Here's how you mount extra paths: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/job_conf.xml.sample_advanced#L600

**Comment by quacksawbones** [comment]
Hi Marius,

I really appreciate you pointing that out to me. While it was a missing piece to the puzzle, it did not fix the problem. But it has allowed me to get to the _root_ of the problem!

Adding in the `singularity_volumes` entry allowed the singularity container to access the information within the Data Library directory (mounting <GALAXY_LIB_DIR>/data1 as a directory in the singularity container). But this directory has links to other folders on our file system (/path/to/user/data1). Galaxy can access it (because the Galaxy and compute nodes have access to that part of shared storage) but because Singularity has no knowledge of /path/to/user/data1, the execution of the tools fail because the links in the container are dead. I have attached a diagram to demonstrate what I mean:

<img width="762" alt="Galaxy_singularity_linking_error" src="https://user-images.githubusercontent.com/1300104/173972488-1a410058-cdb6-49b8-ad64-f9a4e026e19b.png">

This error wasn't seen on our previous Galaxy build because we weren't running Singularity on it. It also wasn't seen and when I was using my Ubuntu test build, link destinations were being mounted in the container, hence why it "started working". And it wasn't seen in the "Library Upload" method, because I wasn't able to access files via SymLink (which is another thing I'm looking at).

We had configured our server so that our Galaxy folders could point to group's "galaxy" directory and allow our users to place links into them as a method of saving space on our cluster and reduce duplication of data files. Are you aware of any way that we can copy the data files to a temporary location (for example, the scratch storage on the compute node) then link to the data files after thenjob has been scheduled but _BEFORE_ Singulary executes? That way I can mount the temporary location in the Singularity containers on whatever compute node the job is sent to, which should make it available for the job to execute but then be deleted afterwards. And would this cause problems with regards to data integrity later on (for example, rerunning a job or downloading the data files)?

Apologies for the long winded infodump. I appreciate your help, Marius,

Darren Cullerne
